### PR TITLE
Tersoff model implementation  

### LINKED ISSUE #162: How can I contribute to the JAX: bond-order potential model?
Hi. I'm SangHyuk Yoo

Thanks for this amazing library.
I am a new user of the JAX library, so I'm not sure how to contribute to the JAX.

Especially, I want to implement bond-order type potential(BOP) models.
I already have a python-based BOP model, and I want to convert it to JAX-MD library format.

Is it okay to fork the JAX-MD and make a pull request after implementing?
If I cannot, please let me know how I can contribute.

Sincerely,
SangHyuk Yoo

### PR DESCRIPTION
**Pull request by SangHyukYoo**
It is a first attempt at including the Tersoff model, one of the bond-order potential models. It has unique bond-order functions that determine the bond strength (resolve #162).

This is the first usage of JAX and JAX-MD for me, so many mistakes and bugs must have occurred. Please let me know what I have to fix. And, I am of course open to all suggestions.

### CODE REVIEW

**Comment by sschoenholz** [comment]
Awesome! Thanks for submitting this. I probably won't have time to review properly for a few days, but at a glance it looks fantastic.

**Comment by ekindogus** [comment]
Hi SangHyuk, Tersoff was definitely a potential type we were interested in having at JAX-MD, thanks for the hard work on this. As Sam said, it looks great at a first glance!

I had a question about what kind of tests you ran to verify that the potential was giving the expected results. I noticed that you added one test to make sure the cubic diamond silicon gets the energy of about -296 (eV?), is that a number that's reported in literature? Are there other tests you've run to verify the potential? (e.g. optimal lattice constant, vacancy energy, average amorphous energy per Si atom etc.)

**Comment by SangHyukYoo** [comment]
Hi. @ekindogus  

The potential energy value, -296 eV, comes from the LAMMPS results when using the same potential file. But, the test was only conducted on my personal PC with RTX 3060. I'm working to have two additional tests that you mentioned. Is it okay to add test functions on the `energy_test.py`?

File: jax_md/energy.py [hunk 1]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+  This function reads multi-element original Tersoff potential parameters
+  from a file.
+  
+  Args:
+    file: File handle for the Tersoff parameters text file.
+
+  Returns:
+    params: ndarray that contains Tersoff parameters.
+  """
+  # start to read files
```
Comment (sschoenholz) [inline jax_md/energy.py:911]: Can you use proper grammer in comments (e.g. capitalization + punctuation)? Thanks!

File: jax_md/energy.py [hunk 2]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+  params = []
+  params_per_line = 17
+  raw_text = file.read().split('\n')
+  file.close()
+
+  # read parameters 
+  # skip if the line has \# or empty
+  # if the number of parameters in one line is less than params_per_line,
+  # additional line is appended to match.
+  is_end = False
```
Comment (sschoenholz) [inline jax_md/energy.py:923]: Perhaps something like this could be a little cleaner: https://pastebin.com/CJgeVRFL

File: jax_md/energy.py [hunk 3]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+  """The cut-off function of the Tersoff potential.
+  Args: 
+    R: A scalar that is the average of inner and outer cutoff radii
+    D: A scalar that is the half of the difference 
+       between inner and outer cutoff radii
+
+  Returns:
+    cut-off values
+  """
+  outer = jnp.where(dr < R+D,
```
Comment (sschoenholz) [inline jax_md/energy.py:980]: Can you add spaces around binary operations, e.g. `R + D` rather than`R+D`.

File: jax_md/energy.py [hunk 4]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+                    0.5*(1.0 - jnp.sin(jnp.pi/2*(dr - R)/D)),
+                    0)    
+  inner = jnp.where(dr < R-D, 1, outer)
+  return inner
+
+def _ters_bij(dRij, dRik, neighbor, mask_ijk,
+              R, D, c, d, h, lam3, beta, n, m) -> Array:
+  """The bond-order term of the Tersoff potential.
+  Args: 
+    R: A scalar that is the average of inner and outer cutoff radii
```
Comment (sschoenholz) [inline jax_md/energy.py:990]: Can you include the rest of the args here (even though they might be self-explanatory, I'm not sure what `neighbor` means for example (which actually seems to be unused). It would be nice to say what shape of input dRij and dRik are expected to be. I assume both are [n, neighbors, dim], is that right?

File: jax_md/energy.py [hunk 5]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+    bond-order values between i and j atoms
+  """
+  drij = space.distance(dRij)
+  drik = space.distance(dRik)
+  
+  # compute g_ijk - angle penalty value
+  costheta = quantity.cosine_angles(dRij)
+  gijk = 1.0 + (c**2 / d**2) - (c**2 / (d**2 + (h - costheta)**2))
+  gijk *= mask_ijk
+  gijk = vmap(lambda x: jax.ops.index_update(x, jnp.diag_indices(x.shape[0]), 0))(gijk)
```
Comment (sschoenholz) [inline jax_md/energy.py:1007]: I think `y = jax.ops.index_update(x, idx, vals)` is being deprecated in favor of `y = x.at[idx].set(vals)`.

File: jax_md/energy.py [hunk 6]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+  fC = _ters_cutoff(drik, R, D)
+
+  # compute zeta without diagonal term
+  tmp = jnp.multiply(gijk, explr3)
+  zeta_ij = jnp.einsum('ik,ijk->ij', fC, tmp)
+  bij = (1 + (beta*zeta_ij)**n)**(-1 / 2 / n)
+  
+  return bij
+
+def _ters_attractive(B: f32, lam2: f32,
```
Comment (sschoenholz) [inline jax_md/energy.py:1027]: Can we keep (close to) the same argument order between `ters_attractive` and `ters_bij`? E.g. can we keep the `dR` and `mask` either at the beginning or the end?

File: jax_md/energy.py [hunk 7]
```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
...
+                                'with Dense neighbor lists.')
+
+  # define compute functions
+  # todo: is neighboring mask necessary? 
+  def compute_fn(R, neighbor, **kwargs):
+    d = partial(displacement, **kwargs)
+    mask = partition.neighbor_list_mask(neighbor, mask_self=True)
+    mask_ijk = mask[:, None, :] * mask[:, :, None]
+
+    if neighbor.format is partition.Dense:
```
Comment (sschoenholz) [inline jax_md/energy.py:1144]: You already check to make sure the neighbor list format is dense above.

File: tests/energy_test.py [hunk 8]
```diff
@@ -726,6 +726,37 @@ def test_eam_neighbor_list(self, num_repetitions, dtype, format):
...
+       [0.25, 0.75, 0.75],
+       [0.50, 0.00, 0.50],
+       [0.75, 0.25, 0.75],
+       [0.50, 0.50, 0.00],
+       [0.75, 0.75, 0.25]], dtype=dtype)
+    atoms = lattice(atoms, 2, lattice_vectors)
+    lattice_vectors *= 2
+    displacement, _ = space.periodic_general(lattice_vectors, fractional_coordinates=True)
+    box_size = np.linalg.det(lattice_vectors) ** (1 / 3)
+    tersoff_parameters = energy.load_lammps_tersoff_parameters(open('tests/data/Si.tersoff', 'r'))
```
Comment (sschoenholz) [inline tests/energy_test.py:750]: Can you keep inside of 80 characters?

**Review by sschoenholz — COMMENTED** [review]
Sorry for the delay, we went away for winter break. This is looking awesome, thanks for the contribution! I have left a few minor comments, but otherwise this looks fantastic and ready to merge.

**Comment by SangHyukYoo** [comment]
Dear @sschoenholz

Thanks for reviewing the codes. I will update the codes with your comments. 
However, the merge is failed because I made some mistakes. I also fixed the problem as soon as possible.

I have a question about fixing the code after making the merge request.
Do I cancel the merge request? Or Do I use the command 'push --force' after fixing the last commit? 

Sincerely,
SangHyuk Yoo

File: jax_md/energy.py (hunk 1 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:911]: Do you mean that I need to change every comment or this line only?

File: jax_md/energy.py (hunk 2 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:923]: Thanks very much. I changed them you suggested.

File: jax_md/energy.py (hunk 4 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:990]: - I wrote more comments to explain the arguments.
- 'Neighbor' is not necessary for now. I deleted it from the arguements.
- Yes. you are right. I also added the dimensions of dRij.

File: jax_md/energy.py (hunk 5 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:1007]: I updated the method you mentioned.

File: jax_md/energy.py (hunk 6 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:1027]: Yes. It is much easier when someone reads the code. Thanks very much.

File: tests/energy_test.py (hunk 8 above)
Comment (SangHyukYoo) [reply tests/energy_test.py:750]: Yes. I will use less characters.

File: jax_md/energy.py (hunk 7 above)
Comment (SangHyukYoo) [reply jax_md/energy.py:1144]: Thanks for your comment. I deleted that part.

**Comment by SangHyukYoo** [comment]
Hi. @sschoenholz 

I'm sorry to reply late. I have updated the codes to merge again.
Please recheck the code.

Sincerely,
SangHyuk Yoo

**Comment by sschoenholz** [comment]
Hi @SangHyukYoo! Thanks for making these changes. Somehow I missed this message until now. I'll take a look and get back to you soon.

**Comment by SangHyukYoo** [comment]
Hi. @sschoenholz 
I'm totally fine. I also replied late too much. Please take your time to review the codes.

**Comment by sschoenholz** [comment]
This looks fantastic, thanks for the contribution!
